vllm/multimodal/processing.py

```diff
@@ -587,26 +588,203 @@ class ProcessorInputs(NamedTuple):
     """Keyword arguments to :meth:`BaseMultiModalProcessor`"""
     prompt_text: str
     mm_data: MultiModalDataDict
-    mm_processor_kwargs: Mapping[str, object]
+    hf_mm_kwargs: Mapping[str, object]
+
+
+class ProcessingCache:
+
+    def __init__(self, capacity: int) -> None:
+        super().__init__()
+
+        # DEBUG: Set to None to disable
+        self.debug_cache_hit_ratio_steps: Optional[int] = None
+
+        self._text_cache = LRUCache[str, BatchFeature](capacity)
+        self._mm_cache = LRUCache[str, BatchFeature](capacity)
+        self._coarse_cache = LRUCache[str, BatchFeature](capacity)
+
+    def maybe_log_cache_stats(self, cache: LRUCache, name: str) -> None:
+        steps = self.debug_cache_hit_ratio_steps
+        if not steps:
+            return
+
+        cache_stats = cache.stat()
+        if cache_stats.total % steps == 0:
+            logger.debug("ProcessingCache: %s.hit_ratio = %.2f", name,
+                         cache_stats.hit_ratio)
+
+    def _iter_bytes_to_hash(self, key: str, obj: object) -> Iterable[bytes]:
+        # Recursive cases
+        if isinstance(obj, (list, tuple)):
+            for elem in obj:
+                yield from self._iter_bytes_to_hash(key, elem)
+            return
+        if isinstance(obj, dict):
+            for k, v in obj.items():
+                yield from self._iter_bytes_to_hash(f"{key}.{k}", v)
+            return
+
+        # Simple cases
+        if isinstance(obj, str):
+            yield key.encode("utf-8")
+            yield obj.encode("utf-8")
+            return
+        if isinstance(obj, bytes):
+            yield key.encode("utf-8")
+            yield obj
+            return
+        if isinstance(obj, Image):
+            yield key.encode("utf-8")
+            yield obj.tobytes()
+            return
+
+        # Convertible to NumPy arrays
+        if isinstance(obj, torch.Tensor):
+            obj = obj.numpy()
+        if isinstance(obj, (int, float)):
+            obj = np.array(obj)
+        if isinstance(obj, np.ndarray):
+            yield key.encode("utf-8")
+            yield obj.tobytes()
+            return
+
+        msg = f"Unable to hash object of type {type(obj)}"
+        raise NotImplementedError(msg)
+
+    def _hash_kwargs(self, **kwargs: object) -> str:
+        hasher = blake3()
+
+        for k, v in kwargs.items():
+            for item_bytes in self._iter_bytes_to_hash(k, v):
+                hasher.update(item_bytes)
+
+        return hasher.hexdigest()
```

Comment:
I'm a bit worried about unintentional hash collisions. Is there a better way to do this?